fix #2293: Properly encode Decimals without any decimal places.

Hultner opened the pull request:
<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->
<!-- See https://pydantic-docs.helpmanual.io/contributing/ for help on Contributing -->

## Change Summary
Adds a custom encoder for Decimals which handles both decimals with decimal places as well as those without.
<!-- Please give a short summary of the changes. -->

## Related issue number
Closes: #2293 
<!-- Are there any issues opened that will be resolved by merging this change? -->

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100% 
  __Tests where at 99% when i ran them before my change, I do have tests covering my changes__
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

Hultner commented:
@PrettyWood @samuelcolvin Any thoughts on this?

samuelcolvin commented on changes/2293-hultner.md:
```diff
@@ -0,0 +1 @@
+fix #2293: Properly encode `Decimal` with, or without any decimal places.
```
```suggestion
Properly encode `Decimal` with, or without any decimal places.
```

Hultner replied:
Fixed

samuelcolvin commented on tests/test_decimal.py:
```diff
@@ -0,0 +1,28 @@
+"""
```
I know the test files could do with cleaning up, but this doesn't require a standalone test file.

Please move into `test_json`.

Hultner replied:
Have done that now :)

samuelcolvin reviewed (commented):
otherwise LGTM.

Hultner commented:
@samuelcolvin Thanks for the feedback! I know that you are busy. 😃 
I moved the test into test_json, right below the time-delta encoder as that's the functionality I saw the most resemblance with. I also removed the prefix from the changelog entry and resolved the merge conflicts agains master which have arose. 

Does it look good now?

samuelcolvin commented on tests/test_json.py:
```diff
@@ -170,6 +170,29 @@ class Config:
+        decimal_places = 0
+        ge = 0
+
+    ObjId = Id
```
This is unnecssary, better to use `condecimal` anyway.

Hultner replied:
You mean the ConstrainedDecimal subclass?  I'm cursion to hear why is condecimal better?

I use the subclasses since typers doesn't like the `id: condecimal(args)` approach, and it's neat with reusable types.

samuelcolvin commented on tests/test_json.py:
```diff
@@ -170,6 +170,29 @@ class Config:
+
+    class Obj(BaseModel):
+        id: ObjId
+        name: str
```
this field is not relevant to the test.

Hultner replied:
Yes, I'll remove it.

samuelcolvin commented on tests/test_json.py:
```diff
@@ -170,6 +170,29 @@ class Config:
+        name: str
+        price: Decimal = Decimal('0.01')
+
+    test_obj = Obj(id=1, name='Test Obj')
+    cycled_obj = Obj.parse_raw(test_obj.json())
+    assert test_obj == cycled_obj
```
This is smart, but not the cleanest way of writing tests. Tests should demonstrate what they're checking explicitly.

Better to do something like

```py
assert Obj(id=1).json() == '{"id": 1, "price": 0.01}'
```

Hultner replied:
I wanted to include this since the problematic behaviour were strictly related to encoding and then parsing an object. 
But I'll change it if you prefer the other way. However it doesn't capture that parsing the object back also works so maybe add a second parse as well?

Using your method that could look like this.
```python
assert Obj(id=1) == Obj.parse_raw('{"id": 1, "price": " 0.01}')
```

Hultner commented:
@samuelcolvin I updated the test and responded to your comments :)

samuelcolvin commented:
This is great, thank you.
